[#64523] Allow 'Scored List' as formula format

EinLama opened the pull request:
# Ticket
https://community.openproject.org/wp/64523

- [x] Allow usage of ScoredList in formula
- [x] During calculation, make sure the selected value is used. To do so, adjust the Strategy used by the Custom Field.
  - [x] Since Hierarchies and Scored Lists share the same strategy: ensure that Hierarchies still behave well w.r.t. rendering and editing them.
  - [x] Ensure the API still works
- [x] Specs
- [x] Recalculate when enabling/disabling the field
- [x] Recalculate when another entry from the list is selected on the overview page

## Out of scope

Recalculating when a scored list is _modified_ in the administration. For example, when a score is changed. For that, there is another work package.

# Screenshots

https://github.com/user-attachments/assets/6ad5717c-01be-40b8-b0ce-dc52ccfa9188


# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

EinLama committed:
[#64523] Allow 'Scored List' as formula format

EinLama committed:
[#64523] Minor spelling corrections

Noted these while reading through ActsAsCustomizable

EinLama committed:
[#64523] Use 'Scored List' score in calculations

EinLama committed:
[#64523] Cache typed_value in HierarchyStrategy

EinLama committed:
[#64523] Hide the score for 'Scored List' display

It was like this before and we will keep it for now.

EinLama committed:
[#64523] API displays formatted_value for Scored Lists

This is the way it behaved before. We will keep it like it was.

EinLama committed:
[#64523] Refactor: cached_ar_object -> cached_typed_value

Ensure that no code has to be duplicated

EinLama committed:
[#64523] Model spec for Scored Lists in Calculated Value

EinLama committed:
[#64523] Sidebar spec for Scored List on project overview

EinLama committed:
[#64523] Use `hierarchical_list?`

toy commented on app/models/custom_value/hierarchy_strategy.rb:
```diff
@@ -29,6 +29,24 @@
+  def typed_value
+    item = cached_typed_value
+
+    item.score.presence || item
```
I see two problems here: it errors if `item` is `nil` and it returns either a number or an AR instance, which will probably lead to confusion

toy commented on app/models/custom_value/hierarchy_strategy.rb:
```diff
@@ -29,6 +29,24 @@
+
+    if item.nil?
+      "#{value} #{I18n.t(:label_not_found)}"
+    elsif item.short.present?
```
Same as
```suggestion
    elsif item.short?
```

toy commented on app/models/custom_value/hierarchy_strategy.rb:
```diff
@@ -47,14 +65,7 @@ def ar_class
-    else
-      item.label
-    end
+    CustomField::Hierarchy::Item.find_by(id: value.to_s)
```
I'm wondering why is `to_s` needed and why `ar_class` is not used:
```suggestion
    ar_class.find_by(id: value)
```

toy commented on lib_static/plugins/acts_as_customizable/lib/acts_as_customizable.rb:
Thanks for proofreading :)

toy commented on spec/features/projects/project_custom_fields/overview_page/sidebar_spec.rb:
```diff
@@ -590,6 +590,51 @@
           end
         end
+
+        context "when one of the values is a scored list", with_flag: { scored_list_custom_fields: true } do
```
As only displaying is tested, I think it is better to just check display of scored list custom field value outside of `with calculated value CFs` and not checking calculated value using it

toy commented on lib/api/v3/utilities/custom_field_injector/link_value_getter.rb:
Maybe better to test addition in this file

toy committed:
separate CustomValue::ScoredListStrategy

toy committed:
use short attribute? reader for item.short.present?

toy committed:
refactor ar_object of CustomValue::ARObjectStrategy subclasses

toy committed:
test hierarchy and scored list strategies

toy committed:
a bit more cleanup of hierarchy strategy

toy committed:
simplify spec for scored list in sidebar

toy committed:
test handling hierarchy and scored list by custom field injector

Kharonus commented on app/models/custom_value/hierarchy_strategy.rb:
```diff
@@ -29,9 +29,19 @@
+    item = cached_typed_value
+
+    if item
+      item.to_s
```
❓  I have limited review possibilities right now, but where is the to_s defined for hierarchical list item values? Before we had the formatting in here (stuff with label and the short in parenthesis). Where would I find it now?

toy replied:
In the model, I saw that it is already there, so doesn't make sense to repeat code in the strategy and added tests to not break it accidentally

Kharonus reviewed (approved):
In general I only superficially touched the strategies, mostly to make it work. If your improvement does what it intends to do, just from code I see no issue with the changes.

Maybe just check the formatting question, so that we do not lose that due to a minor oversight.

ulferts commented on app/models/custom_value/ar_object_strategy.rb:
```diff
@@ -63,11 +57,21 @@ def validate_type_of_value
+        self.memoized_typed_value = ar_object(value)
+      end
+    end
+  end
```
Why this change? It appears to just extract the complete code of one method into another method, leaving the original method to only contain the call to the extracted method.

If it is for [overwriting the method in `HierarchyStrategy`](https://github.com/opf/openproject/pull/20454/files#diff-1f34b79c1121574f3ab7327df44950cf8d57a81ef3ad8394f90118d4b520b19bR32), then it could still call `typed_value` there.

toy replied:
The problem will be with calling `formatted_value` from `ScoredListStrategy`, as `typed_value` is redefined to return the `score`. Maybe the method should be named differently, like `cached_ar_object`, but then `memoized_typed_value` can be confusing, let me try to refactor it

ulferts replied:
The change works for me, I will still struggle when seeing a method which just calls another method but I at least now understand the rational. Maybe a comment would prevent me from boldly changing it when I see it in a couple of months.

ulferts commented on app/models/custom_value/ar_object_strategy.rb:
```diff
@@ -63,11 +57,21 @@ def validate_type_of_value
-  def ar_object(_value)
-    raise NotImplementedError
+  def ar_object(value)
+    ar_class.find_by(id: value)
```
Well spotted change 👍 Quite some methods that can be removed because of it.

ulferts commented on app/models/custom_value/hierarchy_strategy.rb:
```diff
@@ -46,17 +56,6 @@ def ar_class
-    else
-      item.label
-    end
-  end
```
Good change here, that method should definitely not return a string.

ulferts reviewed (changes requested):
Hey @toy , the change itself is fine and the functionality is working as expected.

The reason why I don't approve the PR just now is that I couldn't find a spec that ensures the calculated values work with scored lists. I didn't look around but I would imagine having at least a feature spec that builds a formula from all of the supported cf would be appropriate.

And then there is the one remark about the method extraction but that is of lesser importance.

toy commented:
> The reason why I don't approve the PR just now is that I couldn't find a spec that ensures the calculated values work with scored lists. I didn't look around but I would imagine having at least a feature spec that builds a formula from all of the supported cf would be appropriate.

See the diff for spec/models/acts_as_customizable/calculated_value_spec.rb, test for scored list is added there in 2f3e6b03d22f70cb8c615efc484667402f7bdc5c

toy committed:
move handling of cache to ARObjectStrategy and rework it

Caching is used only in the ARObjectStrategy, so it doesn't need to be
present in base class. Also no need to have separate
memoized_typed_value accessor and cached_typed_value helper.

ulferts commented:
> > The reason why I don't approve the PR just now is that I couldn't find a spec that ensures the calculated values work with scored lists. I didn't look around but I would imagine having at least a feature spec that builds a formula from all of the supported cf would be appropriate.
> 
> See the diff for spec/models/acts_as_customizable/calculated_value_spec.rb, test for scored list is added there in [2f3e6b0](https://github.com/opf/openproject/commit/2f3e6b03d22f70cb8c615efc484667402f7bdc5c)

I admittedly missed that one. Stil would prefer a feature spec in this case. Float and int seem to already be covered so adding either another feature spec for scored lists or condensing them all into one would give me some peace.

toy committed:
use shared_let in shared context "with seeded project custom fields"

toy committed:
dynamically build custom field project and wp sub factories

toy committed:
add missing project and wp sub factories

toy committed:
use multi_ factories in shared context "with seeded project custom fields"

toy committed:
add scored list field to a formula during a feature test

toy committed:
simplify ListStrategy#typed_value

ulferts reviewed (approved):
👍
